### ISSUE TITLE
Suggestions from personal wordlists

### ISSUE THREAD

**Issue by minad** [rrthomas/enchant#403]
I don't recall if this has always been the case, but it seems that Enchant with Hunspell doesn't provide suggestions from personal wordlists. Suggestions are only available before restart. Is this a regression?

See https://github.com/minad/jinx/discussions/217 for the original report.

Thank you!

**Comment by rrthomas** [comment]
I will look into this. I thought it was working, since I use Enchant with the Hunspell back end and a personal wordlist. As you noted, the way that the personal wordlist is fed into Enchant has changed, but in particular with Hunspell personal wordlist words should still appear in suggestions.

**Comment by minad** [comment]
Okay, thanks. I think it is good to leave the suggestion generation to the backend if supported, instead of doing it in Enchant. Hopefully we can get this working again.

**Comment by minad** [comment]
@rrthomas Any news on this issue?

**Comment by rrthomas** [comment]
Sorry, I've not found time to work on this yet. It is still in my inbox!

**Comment by rrthomas** [comment]
I cannot reproduce this bug with current Enchant git, which I shall shortly release as 2.8.3, so others will be able to test (though I don't think I've changed anything relevant since 2.8.2).

I tested as follows:

1. Add the word "fooble" to my personal dictionary.
2. Confirm that if I type "foable" then "fooble" is offered as a correction.
3. Restart Emacs.
4. Confirm again that if I type "foable" then "fooble" is offered as a correction.

**Comment by minad** [comment]
Thanks for checking. At least on 2.8.2 on Debian with Hunspell 1.7.2 I see this issue.

1. Add the word "fooble" to my personal dictionary.
2. Confirm that if I type "foable" then "fooble" is offered as a correction. That's okay.
3. Restart Emacs.
4. If I type "foable" then "fooble" is NOT offered as a correction. This is the original issue - personal suggestions do not survive the restart.

I guess our setups should be pretty similar since I am on Debian testing. So I will be able to confirm this easily as soon as I see 2.8.3.

**Comment by rrthomas** [comment]
For the record, I'm on Ubuntu 24.04, using Hunspell 1.7.2 (as supplied by Ubuntu).

Since you confirmed having the issue with exactly the same steps as me on a very similar system, I also double-checked that I cannot reproduce it, in case I had done something wrong the previous time. I can re-confirm that the procedure above works for me, and suggests "fooble" as a correction after restart.

**Comment by minad** [comment]
Is this bug still present in the newest Enchant? See also https://github.com/minad/jinx/discussions/259.

(I cannot confirm either way since I am on Debian with Enchant 2.8.2 and Hunspell 1.7.2. There the bug is present, see https://github.com/rrthomas/enchant/issues/403#issuecomment-2841102127)

**Comment by rrthomas** [comment]
I couldn't reproduce this in 2.8.3 (see above), and I can't reproduce it in current Enchant 2.8.14.

**Comment by struan-robertson** [comment]
I am the same as @minad for reproducing the bug.
It might be worth noting that if I add "fooble" to my personal dictionary through Jinx, then it is offered as a correction until Emacs restarts.
However, if I add "fooble" manually to my personal dictionary file, it shows as correctly spelled but is not offered as a correction.

I am using Enchant 2.8.14 and Aspell 0.60.8.2 on Gentoo

**Comment by rrthomas** [comment]
Thanks, the detail that you're using Aspell is helpful. I'll test with Aspell and get back to you.

**Comment by rrthomas** [comment]
In fact, I was using Aspell all along, and it works fine. (I was trying to work out why it was correcting the bogus word even after I'd expunged it from my Enchant spelling list, and that was because it was also saved to my Aspell PWL, i.e. `~/.aspell.en.pws`!)

So again, I'll need a procedure to replicate the bug.

**Comment by rrthomas** [comment]
Just to avoid time-wasting, I have double-checked: start Emacs, type "fooble", it's marked as wrong, add it to my PWL, restart Emacs, type "foable", it's marked as wrong, "fooble" is suggestion number 4.

And, just to make really sure, delete "fooble" from both Enchant's PWL and Aspell's PWL, and restart Emacs, and type "fooble" and it is once more marked as wrong. And finally, if I type "foalbe", then "fooble" is no longer offered as a suggestion.

**Comment by minad** [comment]
To be clear, I see the issue with Hunspell. My setup is pretty standard, I am just using the packages from Debian stable.

I think the crucial point is that the word is not offered as candidate when loaded from the en.dic file, but it is offered when the candidate is added to the word list in the current Emacs session. This means there is some difference between the in-memory dictionary and the dictionary loaded from the file?

**Comment by minad** [comment]
FWIW this is what I wrote many months ago:

> If I type "foable" then "fooble" is NOT offered as a correction. This is the original issue - personal suggestions do not survive the restart.

My ~/.config/enchant/enchant.ordering file contains this:

~~~
*:hunspell
~~~

**Comment by rrthomas** [comment]
Sure, I've tried with Hunspell too (using your `enchant.ordering`, thanks!).

**Comment by minad** [comment]
Okay, and you don't see the bug? Neither with Hunspell nor with Aspell? The suggestion survives the Emacs restart in both cases, and it is stored in the ~/.config/enchant/*.dic file and nowhere else? I don't know what else I can provide to reproduce.

**Comment by rrthomas** [comment]
> Okay, and you don't see the bug? Neither with Hunspell nor with Aspell? The suggestion survives the Emacs restart in both cases, and it is stored in the ~/.config/enchant/*.dic file and nowhere else? I don't know what else I can provide to reproduce.

For Aspell the suggestion is stored in `~/.config/enchant/*.dic` and also in `~/.aspell.en.pws`. For Hunspell, it's stored only in `~/.config/enchant/en_GB.dic` (in my case).

Also, twist! I have actually been testing with Ubuntu-supplied Enchant 2.8.2 all along so far. (I had forgotten that Jinx would be ignoring my user-installed Enchant as it's not built with user-specific paths.)

I will force Emacs to load my user-installed latest Enchant and re-test.

BTW, I agree I don't see what else you can do. It's frustrating.

**Comment by rrthomas** [comment]
OK, finally.

I can reproduce the problem with Enchant 2.8.14, with Aspell.

There are still several things that don't make sense: I can't reproduce it with 2.8.2 and Hunspell (like @minad). Or indeed in any other combination.

But at least I can reproduce @struan-robertson's specific situation, and I think I can have a guess as to what's going wrong: it's something to do with the fact that with Enchant 2.8.14, Aspell's `~/.aspell.en.pws` is no longer being updated when I add `fooble` to the wordlist (though the Enchant PWL is being updated).

**Comment by struan-robertson** [comment]
Yes, I can confirm that my `~/.aspell.en.pws` is empty, and that when I add "fooble" to it, it is now offered as a spelling correction!

**Comment by minad** [comment]
@rrthomas Do we use the same Hunspell version? Debian stable has libhunspell-1.7-0:amd64                   1.7.2+really1.7.2-10+b4.

**Comment by rrthomas** [comment]
> [@rrthomas](https://github.com/rrthomas) Do we use the same Hunspell version? Debian stable has libhunspell-1.7-0:amd64                   1.7.2+really1.7.2-10+b4.

Same here, up to the `-10`.

**Comment by minad** [comment]
And you're 100% sure that you're Hunspell and not accidentally Aspell? You can evaluate `(jinx--mod-langs)` to check. No idea what could be the problem on my side 🤷🏾

**Comment by rrthomas** [comment]
> And you're 100% sure that you're Hunspell and not accidentally Aspell? You can evaluate `(jinx--mod-langs)` to check. No idea what could be the problem on my side 🤷🏾

After my less-than-stellar work on this issue so far, a very reasonable question! Also, after looking at the code, I understand why both Hunspell & Aspell would be failing, but not why just one of them would.

Huh, well well well. Can reproduce. (I am on Hunspell.)

That's a relief, because I have a sensible explanation for what's going on now, and a solution.
